src/orion/core/worker/warm_start/multi_task_wrapper.py

```diff
@@ -0,0 +1,245 @@
+"""Wrapper that makes an algorithm warm-starteable by observing trials from different tasks.
+
+The wrapper adds a `task_id` dimension to the search space, and uses the value of `task_id` to
+differentiate between the current "target" experiment (`task_id==0`) and the other experiments
+from the knowledge base (`task_id>0`).
+
+The trials from the knowledge base that can fit within the same space are reused and given a
+different value for a new "task_id" dimension.
+"""
+from __future__ import annotations
+
+import copy
+from collections import defaultdict
+from contextlib import contextmanager
+from logging import getLogger
+from typing import Iterable, TypeVar
+
+from orion.algo.base import BaseAlgorithm
+from orion.algo.space import Categorical, Space
+from orion.core.utils.format_trials import dict_to_trial
+from orion.core.worker.algo_wrappers.algo_wrapper import (
+    AlgoWrapper,
+    _copy_status_and_results,
+)
+from orion.core.worker.warm_start.knowledge_base import ExperimentInfo, KnowledgeBase
+from orion.core.worker.warm_start.warm_starteable import (
+    WarmStarteable,
+)
+from orion.core.worker.trial import Trial
+
+AlgoType = TypeVar("AlgoType", bound=BaseAlgorithm)
+log = getLogger(__file__)
+
+
+class MultiTaskWrapper(AlgoWrapper[AlgoType], WarmStarteable):
+    """Wrapper that makes the algo "multi-task" by adding a task id to the inputs."""
+
+    def __init__(
+        self,
+        space: Space,
+        algorithm: AlgoType,
+    ):
+        log.debug(
+            f"Creating a MultiTaskAlgo wrapper: space {space}, algo config: {algorithm.configuration}"
+        )
+        self.current_task_id: int = 0
+
+        # TODO: IDEA: Use `self.registry` only for the current (target) task, and let `self.algorithm.registry`
+        # contain trials from all tasks, with task IDs added, including the target task?
+
+    def transform(self, trial: Trial) -> Trial:
+        return self._add_task_id(trial, self.current_task_id)
+
+    def reverse_transform(self, trial: Trial) -> Trial:
+        return self._remove_task_id(trial)
+
+    @classmethod
+    def transform_space(cls, space: Space, knowledge_base: KnowledgeBase) -> Space:
+        """Transform the space, so that the algorithm that is passed to the constructor is already
+        transformed.
+        """
+        # TODO: Should we have this dimension be larger than the current number of
+        # experiments in the KB, in case some get added in the future?
+        # TODO: Should the number of tasks here only count the experiments where the spaces are
+        # compatible? (Currently only counts total number of experiments in the KB)
+        # TODO: Could we use a Catagorical over the experiment IDS instead of a Categorical with
```

Comment:
It could help if we don't get consistent order of experiments from one execution to another.